File: two-solid system/FEM/Renumber_Mesh.py
input_file = 'BladeMesh_1.inp'
output_file = 'BladeMesh_1_v2.inp'

first_section_header = 'NODE, NSET=NALL'

offset = int(1e6)

# Initialize a dictionary to store the arrays for each section
data = {}

# Read the lines from the input file
with open(input_file, 'r') as file:
    lines = file.readlines()

# Process the lines and extract the numbers for each section
current_section = None
for line in lines:
    if line.startswith('*'):
        # Start of a new section
        current_section = line.strip().lstrip('*')
        data[current_section] = []
        data[current_section].append(line.rstrip())
    else:
        # Extract numbers from the line
        numbers = line.strip().rstrip(',').split(',')
        if current_section == first_section_header:
            # Change the values of the first section to integers
            numbers[0] = offset+int(numbers[0])
            numbers = [numbers[0]] + [float(num) for num in numbers[1:] if num.strip() != '']
        else:
            numbers = [offset+int(num) for num in numbers if num.strip() != '']
        data[current_section].append(numbers)

# Write the sections with their headers to the output file
# # with open(output_file, 'w') as file:
# #     for section_name, section_data in data.items():
# #         file.write(section_data[0] + '\n')
# #         for row in section_data[1:]:
# #             formatted_row = [str(element) for element in row]
# #             file.write(', '.join(formatted_row) + '\n')

with open(output_file, 'w') as file:
    for section_name, section_data in data.items():
        file.write(section_data[0] + '\n')
        if section_name == first_section_header:
            for row in section_data[1:]:
                formatted_row = [f'{element:.6e}' if isinstance(element, float) else str(element) for element in row]
                file.write(', '.join(formatted_row) + '\n')
        else:
            for row in section_data[1:]:
                formatted_row = [str(element) for element in row]
                file.write(', '.join(formatted_row) + '\n')

File: two-solid system/FEM/RenumberMesh.m
function RenumberMesh(input_file, output_file, first_section_header, offset)

% Read the file
fid = fopen(input_file, 'r');
% read as '%s' strings, split text as newline, do not removes spaces and tabs.
lines = textscan(fid, '%s', 'Delimiter', '\n', 'Whitespace', ''); % lines is 1x1 cell
fclose(fid);
lines = lines{1};

% Prepare output cell array
out_lines = cell(size(lines));

% Flags to know which section we are in
in_node = false;

for i = 1:length(lines)
    line = strtrim(lines{i}); % Remove leading and trailing whitespace from strings

    % Detect section headers
    if startsWith(line, first_section_header, 'IgnoreCase', true) % ignore letter case (uppercase/lowercase)
        in_node = true;
        out_lines{i} = line;
        continue;
    elseif startsWith(line, '*') % Other headers
        in_node = false;
        out_lines{i} = line;
        continue;
    end

    % Process node lines
    if in_node && ~isempty(line)
        nums = sscanf(line, '%f,');
        nums(1) = nums(1) + offset; % only change node index
        out_lines{i} = sprintf('\t%d, %.10e, %.10e, %.10e', nums);
        continue;
    end

    % Process element lines
    if ~in_node && ~isempty(line)
        nums = sscanf(line, '%f,');
        nums = nums + offset; % add offset to all numbers (element ID and node indices)
        out_lines{i} = sprintf('\t%d, ', nums);
        continue;
    end

   
end

% Write the modified file
fid = fopen(output_file, 'w');
for i = 1:length(out_lines)
    fprintf(fid, '%s\n', out_lines{i});
end
fclose(fid);

disp('Done! New file saved as ' + string(output_file));

end